src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+            }
+        } else {
+            if let Some(mut_idx) = ty[0..].find("mut") {
+                mutability.push(true);
+                ty = &ty[mut_idx+4..];
+            } else {
+                mutability.push(false);
+                if ty.contains("const") {
+                    ty = &ty[6..];
+                }
+            }
+        }
+    }
+    let c_ty = c_type(ty);
+    let mutability: Vec<&'static str> = mutability.iter().rev().map(|&m| if m { "*" } else { "const *" }).collect();
+    let result = format!("{} {}", c_ty, mutability.join(" "));
+    result
+}
+
+fn c_type(ty: &str) -> String {
+    match ty {
+        v if is_c_ptr(v) => { rust_ptr_to_c_ptr(v) }
+        "c_void" => "void".to_string(),
+        "boolean" => "boolean".to_string(),
+        "u8" => "unsigned char".to_string(),
+        "c_int" => "int".to_string(),
+        "f32" => "float".to_string(),
+        "c_ulong" => "unsigned long".to_string(),
+        "c_uint" => "unsigned int".to_string(),
+        "usize" => "uintptr_t".to_string(),
+        "JSAMPARRAY_MUT" => "JSAMPARRAY".to_string(),
+        "JSAMPIMAGE_MUT" => "JSAMPIMAGE".to_string(),
+        v => v.to_string(),
+    }
+}
+
+fn c_param(id: &str, ty: &str) -> String {
+    match ty {
+        v => format!("{} {}", c_type(ty), id),
+    }
+}
+
+fn cxx_header() -> String {
+r#"
+    #include <jinclude.h>
+    #include <jpeglib.h>
+    #include <jconfig.h>
+    #include <jconfigint.h>
+    #include <stdexcept>
+
+    struct WrapperError {
+        char* buf;
+        uintptr_t len;
+        uintptr_t cap;
+    };
+
+    struct mozjpeg_rust_wrapper_exception: std::runtime_error {
+        WrapperError error;
+    };
+
+    template <typename T>
+    union WrapperData {
+        T value;
+        WrapperError error;
+        WrapperData(T v) : value(v) {}
+        WrapperData(WrapperError e) : error(e) {}
+    };
+
+    template <>
+    union WrapperData<void> {
+        WrapperError error;
+        WrapperData() {}
+        WrapperData(WrapperError e) : error(e) {}
+    };
+
+    template<typename T>
+    struct WrapperResult {
+        uint8_t discriminant;
+        WrapperData<T> data;
+        WrapperResult(uint8_t d, WrapperData<T> v): discriminant(d), data(v) {}
+        static WrapperResult<T> value(T value) {
+            return WrapperResult(0, WrapperData<T>(value));
+        }
+        static WrapperResult<T> error(WrapperError error) {
+            return WrapperResult(1, WrapperData<T>(error));
+        }
+    };
+
+    template<>
+    struct WrapperResult<void> {
+        uint8_t discriminant;
+        WrapperData<void> data;
+        WrapperResult(uint8_t d, WrapperData<void> v): discriminant(d), data(v) {}
+        static WrapperResult<void> value() {
+            return WrapperResult(0, WrapperData<void>());
+        }
+        static WrapperResult<void> error(WrapperError error) {
+            return WrapperResult(1, WrapperData<void>(error));
+        }
+    };
```

Comment:
Another workaround for lack of void values.

Reply:
I think that all of this could be moved to a helper crate. There is nothing specific about mozjpeg here.